### ISSUE TITLE
image looks resized based on its URL, but fails to resize.

### ISSUE THREAD

**Issue by jmarantz** [apache/incubator-pagespeed-mod#1198]
Consider web site  o p t i g u r a . com (you remove the spaces).  In this snippet I'll use example.com but you have to correct that to repro:

% wget https://www.example.com/img/150x40xexample.png.pagespeed.ic.IW_7v9HkHi.png
--2015-12-02 20:03:37--  https://www.example.com/img/150x40xexample.png.pagespeed.ic.IW_7v9HkHi.png
Resolving www.example.com (www.example.com)... 108.162.202.37, 108.162.201.37
Connecting to www.example.com (www.example.com)|108.162.202.37|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 21700 (21K) [image/png]
Saving to:  150x40xexample.png.pagespeed.ic.IW_7v9HkHi.png 

100%[======================================>] 21,700      --.-K/s   in 0.006s  

2015-12-02 20:03:38 (3.33 MB/s) -  150x40xexample.png.pagespeed.ic.IW_7v9HkHi.png  saved [21700/21700]

% file 150x40xexample.png.pagespeed.ic.IW_7v9HkHi.png 
150x40xexample.png.pagespeed.ic.IW_7v9HkHi.png: PNG image data, 300 x 80, 8-bit/color RGBA, non-interlaced

Note that the file is named 150x40, but is not actually resized; it's 300x80, which is the same as the origin image.

**Comment by huibaolin** [comment]
This image has alpha channel (i.e., has transparent regions). We don't resize images with alpha channel, but we mess up with the URL encoding.

We don't resize images with alpha channel, partially because we used to use OpenCV for resizing and OpenCV didn't resize this format. We don't have this limitation anymore since we've our own resizing. We had tried to extend the support to this format in a refactoring project, which is paused.

I'll revamp the code for resizing alpha channel, which shall fix this bug.

**Comment by Redani** [comment]
Thank you @huibaolin for your reply!
Yes indeed, the logo had a transparent background. But the slides are a .jpg images and they are not resized...

Is there a way to save transparent .png on photoshop to be able to be resized by pagespeed?

**Comment by huibaolin** [comment]
Could you give me the link to the .jpg file which was not resized?

To remove the transparent regions, you can "blend channels" and export the image as a PNG. However, it's probably not a good idea to do that, because it's very difficult to make the new image look the same as the old one.

**Comment by Redani** [comment]
I'm unfortunately not home and don't have my computer, it's about the slides in the home pages, there are 4 or 5 images with 1600px width...

**Comment by huibaolin** [comment]
Did you mean images like this:

<a href="/product/protein-whey/">
<img src="/img/slider/protein-whey.jpg" width="1600" class="img-hero" alt="Protein Whey - Optimum Nutrition"/>
</a>

The original (un-optimized) HTML has marked the width to 1600 pixels, which is the natural width, so is the width of the rewritten image.

**Comment by huibaolin** [comment]
The HTML for the image didn't display correctly last time. Paste it again:

```
<a href="/product/protein-whey/">
<img src="/img/slider/protein-whey.jpg" width="1600" class="img-hero" alt="Protein Whey - Optimum Nutrition"/>
</a>
```

**Comment by Redani** [comment]
> Did you mean images like this:

Yes, i'm talking about that image! as I have a 13" screen, it's resized to 1440px (attached image), whereas pagespeed is supposed to resize it to 1440px!

<img width="832" alt="screen shot 2015-12-03 at 19 29 39" src="https://cloud.githubusercontent.com/assets/328526/11572394/f1c983c4-99f8-11e5-9cd0-166baa39a57f.png">

**Comment by huibaolin** [comment]
Browser displays the image at a size according to the screen/window, for example, you may get one size for 13" screen and another size for 11" screen. This happens after the browsers already receives the image from the server. At server, we don't know what the screen size is so can't resize based on that. What we can do is to resize to the "width" and/or "height" specified in HTML.

If you feel that your website is mostly viewed by a 13" or smaller screen, you can change width from "1600px" to something smaller such as "1440px" as you observed.

**Comment by Redani** [comment]
Ah OK, i see! thanks for your clarification ;)
and is there a work around with jquery/ css so pagespeed will be able to resize the image according to the screen size?

**Comment by dveldhoen** [comment]
Any progress on this? This bug still doesn't seem to be fixed. This is quite a relevant feature to achieve a 100/100 score in Google's PageSpeed Insights. JPEG images get resized perfectly, but PNG or WEBP images dimensions stay original. The newly created URL suggests they are resized though. JPEG images which are converted to WEBP also don't get resized.

**Comment by huibaolin** [comment]
We're added support for resizing PNG image with alpha channel. Do you still see any PNG which is not resized? If so, could you post a link so we can investigate?

We don't optimize WebP images. WebP image is not popular enough yet so has not been a priority.

**Comment by dveldhoen** [comment]
Best situation would be: PNG/JPG -> resized -> convert to WEBP (when browsers supports)
Alternatively: PNG/JPG -> resized

Both are not working for PNG though. JPEG is working fine like I said, only resize. JPEG to WEBP resize is not happening.

See this URL: [www.dannyveldhoen.nl](http://www.dannyveldhoen.nl/)

Same problem occurs with "ModPagespeedEnableFilters resize_rendered_image_dimensions"

I even had to use these filters to even make JPEG working (these are all the filters I am using):
ModPageSpeed on
ModPagespeedRewriteLevel CoreFilters
ModPagespeedDisableFilters convert_jpeg_to_webp
ModPagespeedDisableFilters convert_to_webp_lossless

**Comment by jmarantz** [comment]
Why did you disable convert_jpeg_to_webp and convert_to_webp_lossless ?  If you turns those on, then transcode to webp happens, and so does resizing of original jpegs.

http://www.dannyveldhoen.nl/?PageSpeedFilters=+convert_jpeg_to_webp,+convert_to_webp_lossless,+debug

has images:

http://www.dannyveldhoen.nl/img/250xNxtest.jpg.pagespeed.ic.VXl-Cxp5mU.webp
http://www.dannyveldhoen.nl/img/300xNxdevices.png.pagespeed.ic.fhkCaEbQor.webp

The jpeg looks like it was successfully resized and transcoded to webp.
The png looks like it was successfully transcoded to webp, and it's named as if it were resized to 300px wide, but it does not look like the resizing worked, as its natural size is 1000px.

So I think JPG resize/transcode-to-webp is working fine on your site (if you configure it).  PNG resize/transcode-to-webp is not, which is what this bug tracks.

**Comment by dveldhoen** [comment]
Thanks for your reply. Great that you could verfify that the PNG resizing is indeed not working. JPEG seems fine now. But to return to my original question. Isn't it a high priority to fix this PNG resizing? Since PNG is widly used. This bug prevents aquiring a 100/100 score in Google's Pagespeed Insights. It would be greatly appreciated if this bug could be fixed. Thanks in advance!

**Comment by dveldhoen** [comment]
Is there a timetable available on when this bug will be fixed? The bug is reported in dec 2015.
